test/functional/plugin/lsp_spec.lua

```diff
@@ -6779,4 +6779,76 @@ describe('LSP', function()
       eq(false, exec_lua([[return vim.lsp.is_enabled('foo')]]))
     end)
   end)
+
+  describe('vim.lsp.buf.workspace_diagnostics()', function()
+    --- @param kind lsp.DocumentDiagnosticReportKind
+    local function setup_server(kind)
+      exec_lua(create_server_definition)
+      exec_lua(function()
+        _G.server = _G._create_server({
+          capabilities = {
+            diagnosticProvider = { workspaceDiagnostics = true },
+          },
+          handlers = {
+            ['workspace/diagnostic'] = function(_, _, callback)
+              callback(nil, {
+                items = {
+                  {
+                    kind = kind,
+                    uri = 'file:///fake/uri',
+                    items = {
+                      {
+                        range = {
+                          start = { line = 1, character = 1 },
+                          ['end'] = { line = 1, character = 1 },
+                        },
+                        message = 'Error here!',
+                        severity = 1,
+                      },
+                    },
+                  },
+                },
+              })
+            end,
+          },
+        })
+        vim.lsp.start({ name = 'dummy', cmd = _G.server.cmd })
+        vim.lsp.buf.workspace_diagnostics()
+      end, { kind })
+    end
+
+    it('updates diagnostics obtained with vim.diagnostic.get()', function()
+      setup_server('full')
+
+      retry(nil, nil, function()
+        eq(
+          1,
+          exec_lua(function()
+            return #vim.diagnostic.get()
+          end)
+        )
+      end)
+
+      eq(
+        'Error here!',
+        exec_lua(function()
+          return vim.diagnostic.get()[1].message
+        end)
+      )
+    end)
+
+    it('ignores unchanged diagnostic reports', function()
+      setup_server('unchanged')
+
+      eq(
+        0,
+        exec_lua(function()
+          -- Wait for diagnostics to be processed.
+          vim.uv.sleep(50)
```

Comment:
It doesn't feel right to use `sleep` but if I don't then this test might just become a no-op since diagnostics are likely to be empty before the workspace request is processed.

Reply:
can you use `retry()` to assert the state change?

Reply:
The thing with this test is that I want to assert that diagnostics _are not_ added. If I use `retry` then it will immediately succeed, but I want to make sure that even after processing the request `vim.diagnostic.get` returns nothing.

Reply:
👍  A small wait is the best we can do in those cases.